### ISSUE TITLE
缺少BAM的模块

### ISSUE THREAD

**Issue by sqf041110** [tulingLab/DGMA2-Net#1]
缺少BAM的模块
<img width="692" alt="image" src="https://github.com/tulingLab/DGMA2-Net/assets/174311089/cae54ec2-b245-4272-8ea8-9b6022a55bb8">

**Comment by zhenzhao723** [comment]
确实是缺少，我感觉这里是不是可以换成多头注意力？就是自己写一点多头注意力的代码试试

**Comment by sqf041110** [comment]
> 确实是缺少，我感觉这里是不是可以换成多头注意力？就是自己写一点多头注意力的代码试试

大佬，有没有现成的模块可以用

**Comment by zhenzhao723** [comment]
我是水货，我自己写的。使用resmet18，live_cd的F1大概是84.1。希望对你有帮助
class DEAM(nn.Module):
    def __init__(self, in_d):
        super(DEAM, self).__init__()
        self.conv1 = nn.Sequential(nn.Conv2d(in_d, in_d // 4, 1, 1),
                                   nn.BatchNorm2d(in_d // 4),
                                   nn.ReLU(inplace=True))

        self.conv2 = nn.Sequential(nn.Conv2d(in_d, in_d, 1, 1),
                                   nn.BatchNorm2d(in_d),
                                   nn.ReLU(inplace=True))

    def forward(self, f_map, e_map):
        # 保留尺寸，回复用
        _, _, he, we = f_map.shape
        # 先卷积带c/4,然后分开/8
        e_map_qk = self.conv1(e_map)
        e_map_q, e_map_k = e_map_qk.chunk(2, dim=1)

        # 拉成向量，准备做能量矩阵
        e_map_q = rearrange(e_map_q, 'b c h w -> b c (h w)')
        e_map_k = rearrange(e_map_k, 'b c h w -> b c (h w)').permute(0, 2, 1)
        dots = torch.einsum('bcn,bij->bni', e_map_q, e_map_k)
        attn = dots.softmax(dim=-1)

        # 开始找V
        f_map_v = rearrange(self.conv2(f_map), 'b c h w -> b c (h w)')
        out = torch.einsum('bcn,bij->bin', attn, f_map_v)

        out = rearrange(out, 'b n (h d) -> b n h d', h=he, d=we)

        return out

**Comment by zhenzhao723** [comment]
我是水货，我自己写的。使用resmet18，live_cd的F1大概是84.1。希望对你有帮助。
class DEAM(nn.Module):
    def __init__(self, in_d):
        super(DEAM, self).__init__()
        self.conv1 = nn.Sequential(nn.Conv2d(in_d, in_d // 4, 1, 1),
                                   nn.BatchNorm2d(in_d // 4),
                                   nn.ReLU(inplace=True))

        self.conv2 = nn.Sequential(nn.Conv2d(in_d, in_d, 1, 1),
                                   nn.BatchNorm2d(in_d),
                                   nn.ReLU(inplace=True))

    def forward(self, f_map, e_map):
        # 保留尺寸，回复用
        _, _, he, we = f_map.shape
        # 先卷积带c/4,然后分开/8
        e_map_qk = self.conv1(e_map)
        e_map_q, e_map_k = e_map_qk.chunk(2, dim=1)

        # 拉成向量，准备做能量矩阵
        e_map_q = rearrange(e_map_q, 'b c h w -> b c (h w)')
        e_map_k = rearrange(e_map_k, 'b c h w -> b c (h w)').permute(0, 2, 1)
        dots = torch.einsum('bcn,bij->bni', e_map_q, e_map_k)
        attn = dots.softmax(dim=-1)

        # 开始找V
        f_map_v = rearrange(self.conv2(f_map), 'b c h w -> b c (h w)')
        out = torch.einsum('bcn,bij->bin', attn, f_map_v)

        out = rearrange(out, 'b n (h d) -> b n h d', h=he, d=we)

        return out

**Comment by zhenzhao723** [comment]
我是水货，但是自己写了一个。在live_cd，200个epoch，F1分数84点多。希望对你有帮助
"""
class DEAM(nn.Module):
    def __init__(self, in_d):
        super(DEAM, self).__init__()
        self.conv1 = nn.Sequential(nn.Conv2d(in_d, in_d // 4, 1, 1),
                                   nn.BatchNorm2d(in_d // 4),
                                   nn.ReLU(inplace=True))

        self.conv2 = nn.Sequential(nn.Conv2d(in_d, in_d, 1, 1),
                                   nn.BatchNorm2d(in_d),
                                   nn.ReLU(inplace=True))

    def forward(self, f_map, e_map):
        # 保留尺寸，回复用
        _, _, he, we = f_map.shape
        # 先卷积带c/4,然后分开/8
        e_map_qk = self.conv1(e_map)
        e_map_q, e_map_k = e_map_qk.chunk(2, dim=1)

        # 拉成向量，准备做能量矩阵
        e_map_q = rearrange(e_map_q, 'b c h w -> b c (h w)')
        e_map_k = rearrange(e_map_k, 'b c h w -> b c (h w)').permute(0, 2, 1)
        dots = torch.einsum('bcn,bij->bni', e_map_q, e_map_k)
        attn = dots.softmax(dim=-1)

        # 开始找V
        f_map_v = rearrange(self.conv2(f_map), 'b c h w -> b c (h w)')
        out = torch.einsum('bcn,bij->bin', attn, f_map_v)

        out = rearrange(out, 'b n (h d) -> b n h d', h=he, d=we)

        return out
"""

**Comment by sqf041110** [comment]
这个是加在哪里？还有nn文件是什么？
